File: apps/relay/internal/threads/threads.go
// Package threads folds kind 8109 thread_op events into the relay-signed kind
// 38101 thread_state that clients read.
//
// Why the relay does this at all: thread state is the "blocked on you" queue,
// and a queue every client computes for itself is a queue clients disagree
// about. Folding once, server-side, gives every reader the same answer.
//
// Why it stays auditable: the folded 38101 lists every op id it incorporated in
// `folded_from`. A client that distrusts the projection can fetch those ops and
// recompute it. This is the difference between the relay asserting a state and
// the relay showing its work — and it is what keeps the relay from quietly
// becoming a trusted authority over what a task says.
//
// On a generic relay none of this happens and clients fold locally, reaching the
// same state from the same ops. The projection is an optimisation, never a
// dependency.
package threads

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"

	"github.com/nbd-wtf/go-nostr"

	"github.com/quorum-chat/quorum/apps/relay/internal/protocol"
)

const (
	KindThreadOp    = 8109
	KindThreadState = 38101
	// maxFoldedFrom bounds the audit list. A thread worked on for months would
	// otherwise grow an unbounded tag list inside a replaceable event that is
	// rewritten on every change.
	maxFoldedFrom = 200
)

type Budget struct {
	USD    *float64 `json:"usd,omitempty"`
	Msat   *int64   `json:"msat,omitempty"`
	Tokens *int64   `json:"tokens,omitempty"`
}

// Spent is a cost: what a thread has run up in total, or — as Cost, the same
// shape under the name the `add_spend` op uses — what one turn of work cost.
// They are one type because the total is nothing but the sum of the reports.
type Spent struct {
	TokensIn  *int64   `json:"tokens_in,omitempty"`
	TokensOut *int64   `json:"tokens_out,omitempty"`
	USD       *float64 `json:"usd,omitempty"`
	Msat      *int64   `json:"msat,omitempty"`
}

type Cost = Spent

// State is the body of a kind 38101 event.
type State struct {
	Status     string   `json:"status"`
	Title      string   `json:"title,omitempty"`
	Assignee   *string  `json:"assignee,omitempty"`
	Budget     *Budget  `json:"budget,omitempty"`
	Spent      *Spent   `json:"spent,omitempty"`
	FoldedFrom []string `json:"folded_from"`
	UpdatedAt  int64    `json:"updated_at,omitempty"`
}

// Op is the body of a kind 8109 event. The schema is a discriminated union, so
// only the fields matching `op` are populated.
type Op struct {
	Op       string  `json:"op"`
	Status   string  `json:"status,omitempty"`
	Assignee *string `json:"assignee,omitempty"`
	Title    string  `json:"title,omitempty"`
	Budget   *Budget `json:"budget,omitempty"`
	Cost     *Cost   `json:"cost,omitempty"`
}

// Store is the subset of the event database the projector needs.
type Store interface {
	QueryEvents(ctx context.Context, filter nostr.Filter) (chan *nostr.Event, error)
	ReplaceEvent(ctx context.Context, event *nostr.Event) error
}

// Publisher broadcasts a newly signed event to subscribers.
type Publisher interface {
	BroadcastEvent(event *nostr.Event)
}

type Projector struct {
	index     *protocol.Index
	store     Store
	publish   Publisher
	secretKey string
	pubkey    string

	// One fold at a time per thread. Two ops arriving together would otherwise
	// each read the same prior state and the second would overwrite the first,
	// losing an op that the relay had already acknowledged.
	locks sync.Map // thread id -> *sync.Mutex
}

func New(index *protocol.Index, store Store, publish Publisher, secretKey string) (*Projector, error) {
	pubkey, err := nostr.GetPublicKey(secretKey)
	if err != nil {
		return nil, fmt.Errorf("deriving the relay public key: %w", err)
	}
	return &Projector{
		index:     index,
		store:     store,
		publish:   publish,
		secretKey: secretKey,
		pubkey:    pubkey,
	}, nil
}

func (p *Projector) PublicKey() string { return p.pubkey }

// Fold is a khatru OnEventSaved hook: it reacts to a stored thread_op by
// rewriting the thread's state event.
func (p *Projector) Fold(ctx context.Context, event *nostr.Event) {
	if event.Kind != KindThreadOp {
		return
	}

	thread := protocol.RootEvent(event)
	group := protocol.Group(event)
	if thread == "" || group == "" {
		return
	}

	var op Op
	if err := json.Unmarshal([]byte(event.Content), &op); err != nil {
		return // already rejected by validation; nothing to fold
	}

	mutex, _ := p.locks.LoadOrStore(thread, &sync.Mutex{})
	lock := mutex.(*sync.Mutex)
	lock.Lock()
	defer lock.Unlock()

	state, previous := p.current(ctx, thread)
	if !apply(&state, op) {
		return
	}
	pauseIfExhausted(&state, op)

	at := after(previous)
	state.UpdatedAt = int64(at)
	state.FoldedFrom = append(state.FoldedFrom, event.ID)
	if len(state.FoldedFrom) > maxFoldedFrom {
		state.FoldedFrom = state.FoldedFrom[len(state.FoldedFrom)-maxFoldedFrom:]
	}

	signed, err := p.sign(state, thread, group, at)
	if err != nil {
		return
	}
	if err := p.store.ReplaceEvent(ctx, signed); err != nil {
		return
	}
	p.publish.BroadcastEvent(signed)
}

// after returns the timestamp to sign the next projection with: now, unless a
// projection already exists at or after now.
//
// Two folds inside one second is the normal case — a client publishes
// `set_budget` and `set_status` together, or an agent reports a spend that
// pauses the thread — and the store's addressable replace falls through to
// NIP-01's lowest-id tiebreak when the timestamps are equal. Half the time the
// *older* state wins on a hash, the newer one is silently not stored, and the
// thread is left saying something that was true one op ago.
//
// The relay is the author of these events, so the honest fix is to keep them
// strictly ordered: a projection folded later really is later. The cost is that
// a burst of ops can date the state event a few seconds ahead of the wall
// clock, which no reader minds — NIP-01 orders by `created_at` and that is
// precisely the order being asserted. Client events get no such licence; they
// still face RejectImplausibleTimestamps.
func after(previous nostr.Timestamp) nostr.Timestamp {
	now := nostr.Now()
	if previous >= now {
		return previous + 1
	}
	return now
}

// current reads the thread's existing state, defaulting to a fresh open thread,
// and reports when that state was signed.
func (p *Projector) current(ctx context.Context, thread string) (State, nostr.Timestamp) {
	fresh := State{Status: "open", FoldedFrom: []string{}}

	results, err := p.store.QueryEvents(ctx, nostr.Filter{
		Kinds:   []int{KindThreadState},
		Authors: []string{p.pubkey},
		Tags:    nostr.TagMap{"d": []string{thread}},
		Limit:   1,
	})
	if err != nil {
		return fresh, 0
	}

	// Drain fully rather than breaking early: the store produces on a goroutine
	// that blocks on send, and abandoning the channel would leak it.
	var latest *nostr.Event
	for event := range results {
		if latest == nil || event.CreatedAt > latest.CreatedAt {
			latest = event
		}
	}
	if latest == nil {
		return fresh, 0
	}

	var state State
	if err := json.Unmarshal([]byte(latest.Content), &state); err != nil {
		return fresh, latest.CreatedAt
	}
	if state.FoldedFrom == nil {
		state.FoldedFrom = []string{}
	}
	if state.Status == "" {
		state.Status = "open"
	}
	return state, latest.CreatedAt
}

// apply mutates state by one op, reporting whether anything changed.
func apply(state *State, op Op) bool {
	switch op.Op {
	case "set_status":
		if op.Status == "" || state.Status == op.Status {
			return false
		}
		state.Status = op.Status
	case "assign":
		if same(state.Assignee, op.Assignee) {
			return false
		}
		state.Assignee = op.Assignee
	case "set_title":
		if op.Title == "" || state.Title == op.Title {
			return false
		}
		state.Title = op.Title
	case "set_budget":
		state.Budget = op.Budget
	case "add_spend":
		// Added, never assigned: an op says what a turn cost, not what the
		// total now is. Two agents reporting concurrently must both be counted,
		// and the lock above is what makes that read-add-write safe.
		if op.Cost == nil {
			return false
		}
		state.Spent = addCost(state.Spent, op.Cost)
	default:
		// An op this relay does not know about. Unknown ops are ignored rather
		// than rejected: the protocol's versioning policy says adding one is a
		// MINOR bump, so a newer client must not be broken by an older relay.
		return false
	}
	return true
}

// pauseIfExhausted is the safety valve: a thread that has spent its budget
// stops, and a human is looking at a paused task rather than a bill.
//
// It fires only for the two ops that can change the relationship between the
// spend and the ceiling. A `set_status` must never trigger it, or an exhausted
// thread could never be resumed — the human's "working" would be rewritten back
// to "paused" by the same fold that stored it, and the only visible effect
// would be that the button does nothing. Resuming an exhausted thread is
// therefore allowed and holds until the next spend report; raising the budget
// is what makes it durable, which is the right place for that decision.
//
// A finished thread is left alone. Pausing a `done` task because a late spend
// report crossed a line would un-finish work that is already delivered.
func pauseIfExhausted(state *State, op Op) {
	if op.Op != "add_spend" && op.Op != "set_budget" {
		return
	}
	if state.Status == "paused" || state.Status == "done" {
		return
	}
	if BudgetExhausted(state.Spent, state.Budget) {
		state.Status = "paused"
	}
}

// BudgetExhausted mirrors `checkBudget` in packages/protocol/src/cost.ts, and
// the two must agree: a client folding ops itself on a channel this relay
// cannot read has to reach the same verdict, or the same thread is paused in
// one place and running in another.
//
// Three rules, each of them a decision: a budget with nothing set is not a
// ceiling of zero; any stated dimension is enough, because whoever set two
// ceilings meant both; and `>=` is exhausted, which makes `set_budget {usd: 0}`
// an immediate freeze using a capability that already exists.
func BudgetExhausted(spent *Spent, budget *Budget) bool {
	if budget == nil {
		return false
	}
	if budget.Tokens != nil && tokensSpent(spent) >= *budget.Tokens {
		return true
	}
	if budget.USD != nil && floatValue(costUSD(spent)) >= *budget.USD {
		return true
	}
	if budget.Msat != nil && intValue(costMsat(spent)) >= *budget.Msat {
		return true
	}
	return false
}

// tokensSpent adds both halves of the spend, because a budget states tokens as
// one number. A thread capped at 40,000 that has spent 39,000 in and 38,000 out
// is nearly twice over and looks comfortable from either column alone.
func tokensSpent(spent *Spent) int64 {
	if spent == nil {
		return 0
	}
	return intValue(spent.TokensIn) + intValue(spent.TokensOut)
}

// addCost sums two costs dimension by dimension, keeping absent absent.
//
// `{}` and `{"usd": 0}` are different claims — "nobody said" and "it was free"
// — and a thread priced in tokens should not grow a `usd` total that reads as
// having been costed.
func addCost(a, b *Cost) *Cost {
	if a == nil && b == nil {
		return nil
	}
	if a == nil {
		a = &Cost{}
	}
	if b == nil {
		b = &Cost{}
	}
	return &Cost{
		TokensIn:  addInts(a.TokensIn, b.TokensIn),
		TokensOut: addInts(a.TokensOut, b.TokensOut),
		USD:       addFloats(a.USD, b.USD),
		Msat:      addInts(a.Msat, b.Msat),
	}
}

func addInts(a, b *int64) *int64 {
	if a == nil && b == nil {
		return nil
	}
	sum := intValue(a) + intValue(b)
	return &sum
}

func addFloats(a, b *float64) *float64 {
	if a == nil && b == nil {
		return nil
	}
	sum := floatValue(a) + floatValue(b)
	return &sum
}

func costUSD(spent *Spent) *float64 {
	if spent == nil {
		return nil
	}
	return spent.USD
}

func costMsat(spent *Spent) *int64 {
	if spent == nil {
		return nil
	}
	return spent.Msat
}

func intValue(v *int64) int64 {
	if v == nil {
		return 0
	}
	return *v
}

func floatValue(v *float64) float64 {
	if v == nil {
		return 0
	}
	return *v
}

func same(a, b *string) bool {
	switch {
	case a == nil && b == nil:
		return true
	case a == nil || b == nil:
		return false
	default:
		return *a == *b
	}
}

func (p *Projector) sign(state State, thread, group string, at nostr.Timestamp) (*nostr.Event, error) {
	content, err := json.Marshal(state)
	if err != nil {
		return nil, err
	}

	event := &nostr.Event{
		Kind:      KindThreadState,
		CreatedAt: at,
		Content:   string(content),
		Tags: nostr.Tags{
			{"d", thread},
			{"h", group},
			{"alt", altFor(state)},
		},
	}
	if err := event.Sign(p.secretKey); err != nil {
		return nil, err
	}
	return event, nil
}

// altFor writes the NIP-31 fallback the protocol requires on every Quorum kind.
//
// Kept generic on purpose. This tag is not encrypted even when the channel is,
// so a title or an assignee's key here would leak past the encryption sitting
// beside it. "a thread is blocked" is all a reader without the body is owed.
func altFor(state State) string {
	return fmt.Sprintf("thread state: %s", state.Status)
}

File: packages/sdk/src/threads.ts
/**
 * The task list: every thread in a workspace, with its state, and with the
 * relay's projection of that state checked rather than believed.
 *
 * Pure, and in the SDK for the same reason `inbox()` is: the console and the
 * reference client both show "what is being worked on", and two implementations
 * of a status projection drift in ways nobody notices until a task is `done` on
 * one screen and `blocked` on another.
 *
 * ## Why this recomputes what the relay already computed
 *
 * Kind 38101 is signed by the relay, and the relay is the one party with both
 * the motive and the position to misstate it. `folded_from` is what makes that
 * checkable — it lists the op ids the relay folded, **in the order it folded
 * them** — so a client holding those ops can replay them and compare. That list
 * has been in the projection since M2 and nothing has ever read it. This does.
 *
 * The order matters and is the reason `folded_from` has to exist at all. The
 * relay folds ops as they arrive, which is not an order any client can
 * reconstruct: `(created_at, id)` is a different sequence, and folding
 * `set_status: done` before or after `set_status: blocked` gives different
 * answers. Without the list, "the relay is lying" and "the relay saw these in a
 * different order" would be indistinguishable, and a check that cannot tell
 * those apart is a check nobody can act on.
 *
 * So each thread reports one of four verdicts:
 *
 * - `local` — no projection at all. That is the generic-relay case, and it is
 *   supposed to work: we fold every op ourselves and lose nothing but the
 *   cross-check.
 * - `agrees` — replaying `folded_from` reproduces the body the relay signed.
 * - `unverifiable` — the relay folded ops we do not hold, so we cannot replay
 *   it. Its state is used, and labelled. Ordinary rather than sinister: a long
 *   thread's early ops fall outside a client's backfill window.
 * - `disagrees` — we hold every op it claims to have folded, replayed them in
 *   its own order, and got something else. There is no innocent version of
 *   this. Our fold is what gets displayed.
 *
 * A forged 38101 on a generic relay — where nothing stops a member signing one
 * — therefore produces a warning rather than a lie, provided the ops are there
 * to check it against.
 *
 * ## What this does not do
 *
 * It does not check whether the author of an op was *allowed* to publish it.
 * Authority is the relay's job at write time (`RequireGrantToSetBudget`) and the
 * resource's job at use time. A client folding ops it was served is computing
 * what the relay would compute, not deciding who may do what — and on a generic
 * relay, which enforces nothing, no amount of local folding would make the
 * answer safe.
 */

import {
  Kinds,
  TagName,
  ThreadOpBody,
  ThreadStateBody,
  addCost,
  checkBudget,
  tagValue,
  type Budget,
  type Cost,
  type EventRef,
  type NostrEvent,
  type ThreadStatus,
} from '@quorum/protocol'
import type { PublishOptions } from './publish.ts'

/** What replaying `folded_from` said about the relay's projection. */
export type ProjectionCheck =
  | { verdict: 'local' }
  | { verdict: 'agrees' }
  | { verdict: 'unverifiable'; missing: number }
  | { verdict: 'disagrees'; fields: string[] }

export interface Thread {
  /** The thread id, which is the id of its immutable kind 11 root. */
  id: string
  root: NostrEvent
  title: string
  status: ThreadStatus
  assignee?: string
  budget?: Budget
  spent?: Cost

  /** The relay's kind 38101, if one has been served. */
  projection?: NostrEvent
  check: ProjectionCheck
  /** Ops the projection does not claim to have folded. Applied on top of it. */
  unfolded: NostrEvent[]

  /** Kind 1111 comments in this thread. */
  replies: number
  /** Newest `created_at` anywhere in the thread, including the root. */
  lastActivity: number
  /** Distinct pubkeys that have published in it, the root author first. */
  participants: string[]
}

export interface ThreadsOptions {
  /** Keep only these statuses. */
  status?: readonly ThreadStatus[]
  /** Keep only threads assigned to this pubkey. */
  assignee?: string
}

/**
 * Every thread in `events`, newest activity first.
 *
 * Threads with no root are skipped rather than synthesised. A comment whose
 * root we have not fetched is a hole in the backfill, and inventing a thread
 * from it would put a task on the screen with no title, no author and no state
 * — which reads as a bug in the workspace rather than a gap in the query.
 */
export function threads(events: readonly NostrEvent[], options: ThreadsOptions = {}): Thread[] {
  const roots = events.filter((e) => e.kind === Kinds.Thread)
  const byThread = groupByRoot(events)

  const out: Thread[] = []
  for (const root of roots) {
    const inThread = byThread.get(root.id) ?? []
    const thread = derive(root, inThread)

    if (options.status && !options.status.includes(thread.status)) continue
    if (options.assignee && thread.assignee !== options.assignee) continue
    out.push(thread)
  }

  return out.sort((a, b) => b.lastActivity - a.lastActivity || (a.id < b.id ? -1 : 1))
}

/** One thread, by id, or `undefined` if its root is not in `events`. */
export function thread(events: readonly NostrEvent[], id: string): Thread | undefined {
  return threads(events).find((t) => t.id === id)
}

/**
 * A thread's mutable state, without needing its root.
 *
 * {@link threads} refuses to synthesise a thread it has no kind 11 for, and that
 * is right for a task list: a row with no title and no author reads as a bug.
 * But an agent about to spend money needs one question answered — *how much is
 * left* — and it is already inside the thread, holding its id. Making it fetch
 * a root it will not look at, in order to find out whether it may work, would be
 * a query it pays for on every action and a failure mode where the budget check
 * silently passes because the backfill window missed a two-year-old root.
 *
 * Takes whatever `events` contain: the 38101 keyed by `d` = `id`, and the 8109
 * ops tagged `E` = `id`. Anything else is ignored. With neither, the answer is
 * an open thread with no budget — which is what an unmanaged thread is.
 */
export function threadState(events: readonly NostrEvent[], id: string): Folded {
  const inThread = (groupByRoot(events).get(id) ?? []).filter(
    (e) => e.kind === Kinds.ThreadOp || e.kind === Kinds.ThreadState,
  )
  const ops = inThread
    .filter((e) => e.kind === Kinds.ThreadOp)
    .sort((a, b) => a.created_at - b.created_at || (a.id < b.id ? -1 : 1))

  const projection = latestProjection(inThread, id)
  const body = projection && ThreadStateBody.safeParse(parse(projection.content))
  const claimed = body?.success ? body.data : undefined

  const { base, folded } = reconcile(claimed, ops)
  return foldOps(
    ops.filter((e) => !folded.has(e.id)),
    base,
  )
}

/**
 * Ask for a change to a thread's state.
 *
 * A request and not a write: the op is a kind 8109 the author signs, and what
 * it means is decided by whoever folds it. The reference relay folds it into
 * 38101 and signs that; a generic relay folds nothing and every reader folds it
 * for themselves via {@link threads}. Both give the same answer, which is the
 * property that lets this system run on a relay that has never heard of Quorum.
 *
 * The `E` tag is how the op finds its thread, so the root must be passed whole
 * rather than by id: NIP-22 wants `E`/`K`/`P` together, and a root scope with
 * the kind or author missing is not a comment on anything.
 *
 * Authority is not checked here and could not be. `set_budget` is gated by the
 * relay (`RequireGrantToSetBudget`) because a spending ceiling is authority
 * rather than coordination; the rest are coordination, and anyone in the group
 * may publish them. A client that pre-checked would only be guessing at a
 * decision the relay makes.
 *
 * `add_spend` is deliberately ungated, and it is self-reported. An agent can
 * under-report what it cost, and nothing here stops it — what the op buys is
 * that the total is *stated* rather than estimated, so it is auditable by
 * replay and works unchanged on a channel the relay cannot read. An agent that
 * would lie about its own spend is an agent that should not hold a budget,
 * which is what capabilities and revocation are for.
 */
export function threadOp(root: EventRef, op: ThreadOpBody): PublishOptions {
  return { kind: Kinds.ThreadOp, thread: root, body: op }
}

/** The mutable half of a thread: what ops fold into. */
export interface Folded {
  status: ThreadStatus
  title?: string
  assignee?: string
  budget?: Budget
  spent?: Cost
}

/**
 * Apply thread ops to a state, in the order given.
 *
 * The same rules as `apply()` in the relay's projector, including that an op
 * changing nothing is not an op: the relay does not record those in
 * `folded_from`, so a replay that applied them would drift from the list it is
 * checking. Unknown ops are ignored rather than rejected — adding one is a
 * MINOR version bump, and an older client must not be broken by a newer one.
 */
export function foldOps(ops: readonly NostrEvent[], from?: Folded): Folded {
  const state: Folded = { status: 'open', ...from }

  for (const event of ops) {
    const parsed = ThreadOpBody.safeParse(parse(event.content))
    if (!parsed.success) continue
    const op = parsed.data

    switch (op.op) {
      case 'set_status':
        state.status = op.status
        break
      case 'assign':
        state.assignee = op.assignee ?? undefined
        break
      case 'set_title':
        state.title = op.title
        break
      case 'set_budget':
        state.budget = op.budget
        break
      case 'add_spend':
        state.spent = addCost(state.spent, op.cost)
        break
    }

    pauseIfExhausted(state, op.op)
  }

  return state
}

/**
 * The budget's teeth, and the reason they are in the fold rather than beside
 * it: the relay pauses an exhausted thread as part of folding, so a client that
 * replayed `folded_from` without this rule would reproduce a different status
 * and report a truthful relay as `disagrees`. The check that catches a lying
 * relay only works if both sides compute the same thing.
 *
 * Fires only on the two ops that change the spend-to-ceiling relationship. A
 * `set_status` must never trigger it, or a human could never resume an
 * exhausted thread — their `working` would be rewritten to `paused` by the
 * same fold that stored it. A `done` thread is left alone, because a late
 * spend report must not un-finish delivered work.
 */
function pauseIfExhausted(state: Folded, op: ThreadOpBody['op']): void {
  if (op !== 'add_spend' && op !== 'set_budget') return
  if (state.status === 'paused' || state.status === 'done') return
  if (checkBudget(state.spent, state.budget).exhausted) state.status = 'paused'
}

function derive(root: NostrEvent, inThread: readonly NostrEvent[]): Thread {
  const ops = inThread
    .filter((e) => e.kind === Kinds.ThreadOp)
    .sort((a, b) => a.created_at - b.created_at || (a.id < b.id ? -1 : 1))
  const projection = latestProjection(inThread, root.id)

  const body = projection && ThreadStateBody.safeParse(parse(projection.content))
  const claimed = body?.success ? body.data : undefined

  const { check, base, folded } = reconcile(claimed, ops)
  // Whatever the relay has not folded yet goes on top. A thread op published a
  // moment ago must show up immediately; waiting for the projection to catch up
  // would make the client feel broken in the one place it is fastest.
  const unfolded = ops.filter((e) => !folded.has(e.id))
  const state = foldOps(unfolded, base)

  const activity = [root, ...inThread]
  return {
    id: root.id,
    root,
    title: state.title ?? tagValue(root.tags, TagName.Title) ?? firstLine(root.content),
    status: state.status,
    assignee: state.assignee,
    budget: state.budget,
    spent: state.spent,
    projection,
    check,
    unfolded,
    replies: inThread.filter((e) => e.kind === Kinds.Comment).length,
    lastActivity: Math.max(...activity.map((e) => e.created_at)),
    participants: [...new Set([root.pubkey, ...activity.map((e) => e.pubkey)])],
  }
}

/**
 * Replay what the relay says it folded, and decide whether to believe the rest.
 *
 * `base` is the state to apply unfolded ops on top of: our own replay when we
 * could do one, the relay's word when we could not.
 */
function reconcile(
  claimed: ThreadStateBody | undefined,
  ops: readonly NostrEvent[],
): { check: ProjectionCheck; base: Folded | undefined; folded: Set<string> } {
  if (!claimed) return { check: { verdict: 'local' }, base: undefined, folded: new Set() }

  const byId = new Map(ops.map((e) => [e.id, e]))
  const replayed = claimed.folded_from.map((id) => byId.get(id))
  const missing = replayed.filter((e) => e === undefined).length
  const folded = new Set(claimed.folded_from)

  if (missing > 0) {
    return { check: { verdict: 'unverifiable', missing }, base: asFolded(claimed), folded }
  }

  const ours = foldOps(replayed as NostrEvent[])
  const fields = differences(ours, claimed)
  return {
    check: fields.length === 0 ? { verdict: 'agrees' } : { verdict: 'disagrees', fields },
    base: ours,
    folded,
  }
}

/**
 * Which fields the relay and the replay disagree about.
 *
 * `updated_at` is excluded deliberately: it is the relay's own clock at fold
 * time, so no replay can reproduce it and comparing it would make every
 * projection look forged.
 *
 * `spent` is compared, and until M8 it could not be: nothing folded into it, so
 * a replay had nothing to say. Now every increment arrives as an `add_spend` op
 * in `folded_from`, which makes the total the one number in the projection a
 * client can check by addition — and a relay inflating what a thread cost is a
 * more tempting lie than a relay misstating its title.
 */
function differences(ours: Folded, theirs: ThreadStateBody): string[] {
  const out: string[] = []
  if (ours.status !== theirs.status) out.push('status')
  if ((ours.title ?? '') !== (theirs.title ?? '')) out.push('title')
  if ((ours.assignee ?? '') !== (theirs.assignee ?? '')) out.push('assignee')
  if (JSON.stringify(ours.budget ?? null) !== JSON.stringify(theirs.budget ?? null)) {
    out.push('budget')
  }
  if (!sameCost(ours.spent, theirs.spent)) out.push('spent')
  return out
}

/**
 * Dimension by dimension rather than by serialising, because the two sides are
 * built by different languages: an absent dimension and a zero must compare
 * unequal, but key order must not matter.
 */
function sameCost(ours: Cost | undefined, theirs: Cost | undefined): boolean {
  for (const key of ['tokens_in', 'tokens_out', 'usd', 'msat'] as const) {
    if ((ours?.[key] ?? null) !== (theirs?.[key] ?? null)) return false
  }
  return true
}

function asFolded(state: ThreadStateBody): Folded {
  return {
    status: state.status,
    title: state.title,
    assignee: state.assignee ?? undefined,
    budget: state.budget,
    spent: state.spent,
  }
}

/**
 * The current 38101 for a thread.
 *
 * NIP-01's addressable rule — newest `created_at`, lowest id on a tie — and no
 * check of who signed it. On our relay nobody else can: `RejectRelaySignedForgeries`
 * refuses a 38101 from anyone but the relay. On a generic relay anyone can, and
 * that is exactly the case the `folded_from` replay is here to catch. Filtering
 * by a pubkey the client would have to be told out of band would trade a
 * detectable problem for an undetectable one.
 */
function latestProjection(inThread: readonly NostrEvent[], id: string): NostrEvent | undefined {
  return inThread
    .filter((e) => e.kind === Kinds.ThreadState && tagValue(e.tags, TagName.Identifier) === id)
    .sort((a, b) => b.created_at - a.created_at || (a.id < b.id ? -1 : 1))[0]
}

/**
 * Index every event by the thread it belongs to.
 *
 * Threads are identified by the NIP-22 root scope `E` for conversation, and by
 * `d` for the addressable state, which carries no `E` — it is not *in* the
 * thread, it is *about* it.
 */
function groupByRoot(events: readonly NostrEvent[]): Map<string, NostrEvent[]> {
  const out = new Map<string, NostrEvent[]>()
  for (const event of events) {
    const id =
      event.kind === Kinds.ThreadState
        ? tagValue(event.tags, TagName.Identifier)
        : tagValue(event.tags, TagName.RootEvent)
    if (!id) continue
    const list = out.get(id)
    if (list) list.push(event)
    else out.set(id, [event])
  }
  return out
}

function firstLine(content: string): string {
  const line = content.split('\n', 1)[0] ?? ''
  return line.length > 80 ? `${line.slice(0, 79)}…` : line
}

function parse(content: string): unknown {
  try {
    return JSON.parse(content)
  } catch {
    return undefined
  }
}
